File: src/accounts/utilities.cpp
/***************************************************************************
**
** Copyright (c) 2012, Tarek Galal <tarek@wazapp.im>
**
** This file is part of Wazapp, an IM application for Meego Harmattan
** platform that allows communication with Whatsapp users.
**
** Wazapp is free software: you can redistribute it and/or modify it under
** the terms of the GNU General Public License as published by the
** Free Software Foundation, either version 3 of the License, or
** (at your option) any later version.
**
** Wazapp is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
** See the GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with Wazapp. If not, see http://www.gnu.org/licenses/.
**
****************************************************************************/
#include "utilities.h";
namespace Utilities{
    QSystemNetworkInfo ninfo;
    QSystemDeviceInfo dinfo;

    QString getImsi(){return dinfo.imsi();}
    QString getImei(){return dinfo.imei();}
    QString getCountryCode(){return ninfo.homeMobileCountryCode();}
    QString getMcc(){return ninfo.currentMobileCountryCode();}
    QString getMnc(){
        QString mnc = ninfo.currentMobileNetworkCode();

        while(mnc.length() < 3){
            mnc = "0"+mnc;
        }
        return mnc;

    }



    QString reverseString(QString str)
    {
        QString res ="";
        for(int i = str.length()-1; i>=0; i--){
            res = res+str.at(i);
        }

        return res;


      char *p = str.toAscii().data();
      char *q = p;

      while(q && *q) ++q;
      for(--q; p < q; ++p, --q)
        *p = *p ^ *q,
        *q = *p ^ *q,
        *p = *p ^ *q;

      str = QString(p);

      return str;
    }



    std::string itoa(int value, unsigned int base) {

        const char digitMap[] = "0123456789abcdef";

        std::string buf;
        // Guard:
        if (base == 0 || base > 16) {

            // Error: may add more trace/log output here
            return buf;
        }
        // Take care of negative int:
        std::string sign;
        int _value = value;
        // Check for case when input is zero:
        if (_value == 0) return "0";
        if (value < 0) {
            _value = -value;
            sign = "-";
        }

        // Translating number to string with base:

        for (int i = 30; _value && i ; --i) {

            buf = digitMap[ _value % base ] + buf;

            _value /= base;

        }

        return sign.append(buf);

    }



    QString getChatPassword(){


        QString imei = getImei();
        QString buffer_str = reverseString(imei);

        qDebug()<<buffer_str;

        QCryptographicHash digest(QCryptographicHash::Md5);
        digest.reset();
        digest.addData(buffer_str.toAscii());


        QByteArray bytes = digest.result();


        buffer_str = "";

        for(int i =0; i<bytes.length(); i++){
            int tmp = bytes.at(i)+128;
            int c = (tmp >> 8) & 0xff;
            int f = tmp & 0xff;

            buffer_str=buffer_str + QString::fromStdString(itoa(f,16));

        }


        return buffer_str;

    }

}

File: src/client/utilities.py
'''
Copyright (c) 2012, Tarek Galal <tarek@wazapp.im>

This file is part of Wazapp, an IM application for Meego Harmattan platform that
allows communication with Whatsapp users

Wazapp is free software: you can redistribute it and/or modify it under the 
terms of the GNU General Public License as published by the Free Software 
Foundation, either version 2 of the License, or (at your option) any later 
version.

Wazapp is distributed in the hope that it will be useful, but WITHOUT ANY 
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A 
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with 
Wazapp. If not, see http://www.gnu.org/licenses/.
'''
import md5
import string

from QtMobility.SystemInfo import QSystemDeviceInfo,QSystemNetworkInfo
class Utilities():

	debug_mode = 1;
	
	waversion = "0.9.15.1"
	

	
	@staticmethod
	def getImsi():
		#return "000000000000000"
		dev_info = QSystemDeviceInfo();
		return dev_info.imsi();
		
	@staticmethod	
	def getProfile():
		dev_info = QSystemDeviceInfo();
		return dev_info.currentProfile();
	
	@staticmethod
	def getCountryCode():
		net_info = QSystemNetworkInfo();
		return net_info.homeMobileCountryCode();
	
	@staticmethod
	def getMcc():
		net_info = QSystemNetworkInfo();
		return net_info.currentMobileCountryCode();
	
	@staticmethod
	def getMnc():
		net_info = QSystemNetworkInfo();
		return net_info.currentMobileNetworkCode();
		
	@staticmethod
	def getImei():
		dev_info = QSystemDeviceInfo();
		return dev_info.imei();
	
	@staticmethod
	def hashCode(string):
		h = 0;
		off =0;
		for i  in range (0,len(string)):
			h = 31*h + ord(string[off]);
			off+=1;

		return h;



	@staticmethod
	def decodeString(url):
		res = "";
		
		for char in url:
			ored = char ^ 0x13
			res = res+chr(ored);
		return res;
	
	@staticmethod
	def encodeString(string):
		res = [];
		for char in string:
			res.append(ord(char))
		return res;
	
	@staticmethod
	def byteArrayToStr(bytearray):
		res = "";
		for b in bytearray:
			res = res+chr(b);
		
		return res;

			

	@staticmethod
	def str( number, radix ):
	   """str( number, radix ) -- reverse function to int(str,radix) and long(str,radix)"""

	   if not 2 <= radix <= 36:
	      raise ValueError, "radix must be in 2..36"

	   abc = string.digits + string.letters

	   result = ''

	   if number < 0:
	      number = -number
	      sign = '-'
	   else:
	      sign = ''

	   while True:
	      number, rdigit = divmod( number, radix )
	      result = abc[rdigit] + result
	      if number == 0:
		 return sign + result

	@staticmethod	
	def getChatPassword():
		imei = Utilities.getImei();
		
		
		buffer_str = imei[::-1];
		digest = S40MD5Digest();
		digest.reset();
		digest.update(buffer_str);
		bytes = digest.digest();
		buffer_str = ""
		for b in bytes:
			tmp = b+128;
			c = (tmp >> 8) & 0xff
			f = tmp & 0xff
			buffer_str+=Utilities.str(f,16);
		return buffer_str;
	


class ByteArray():
	def __init__(self,size=0):
		self.size = size;
		self.buf = bytearray(size);	
	
	def toByteArray(self):
		res = ByteArray();
		for b in self.buf:
			res.buf.append(b);
			
		return res;

	def reset(self):
		self.buf = bytearray(self.size);
		
	def getBuffer(self):
		return self.buf
		
	def read(self):
		return self.buf.pop(0);
		
	def read2(self,b,off,length):
		'''reads into a buffer'''
		if off < 0 or length < 0 or (off+length)>len(b):
			raise Exception("Out of bounds");
		
		if length == 0:
			return 0;
		
		if b is None:
			raise Exception("XNull pointerX");
		
		count = 0;
		
		while count < length:
			
			#self.read();
			#print "OKIIIIIIIIIIII";
			#exit();
			b[off+count]=self.read();
			count= count+1;
		
	
		return count;
		
	def write(self,data):
		if type(data) is int:
			self.writeInt(data);
		elif type(data) is chr:
			self.buf.append(ord(data));
		elif type(data) is str:
			self.writeString(data);
		elif type(data) is bytearray:
			self.writeByteArray(data);
		else:
			raise Exception("Unsupported datatype "+str(type(data)));
			
	def writeByteArray(self,b):
		for i in b:
			self.buf.append(i);
	
	def writeInt(self,integer):
		self.buf.append(integer);
	
	def writeString(self,string):
		for c in string:
			self.writeChar(c);
			
	def writeChar(self,char):
		self.buf.append(ord(char))
		

class S40MD5Digest():
	m = None;
	def __init__(self):
		self.m= md5.new()
		
	def update(self,string):
		#Utilities.debug("update digestion");
		self.m.update(str(string));
		
	def reset(self):	
		self.m = md5.new();
		
	def digest(self):
		#res = self.m.digest();
		#return res;
		arr = bytearray(128);
		res = 0;
		res = self.m.digest();
		resArr = bytearray(res);
		
		return resArr;
